inline/src/inlines/mod.rs

```diff
@@ -1,231 +1,285 @@
-    Parentheses(PlainContent),
+    Math(Math),
 
     /// Content of a TextGroup `[]`.
-    TextGroup(NestedContent),
+    TextGroup(TextGroup),
 
     /// Unimarkup attributes for some content.
-    Attributes(NestedContent),
+    Attributes(Attributes),
 
     /// Alias substitution ( i.e. `::heart::`).
-    Substitution(NestedContent),
+    Substitution(Substitution),
+
+    /// Wrapper without any special formatting for multiple other [`Inline`]s.
+    ///
+    /// [`Inline`]: self::Inline
+    Multiple(Multiple),
+
+    /// Content inside a pair of parenthesis `()`.
+    Parentheses(Parentheses),
+
+    /// Verbatim (monospaced) content.
+    Verbatim(Verbatim),
 
     /// Explicit newline.
-    Newline(PlainContent),
+    Newline(Newline),
 
     /// Explicit whitespace.
-    Whitespace(PlainContent),
+    Whitespace(Whitespace),
 
     /// End of line (regular newline)
-    EndOfLine(PlainContent),
+    EndOfLine(EndOfLine),
 
     /// Plain text without any formatting.
-    Plain(PlainContent),
-
-    /// Wrapper without any special formatting for multiple other [`Inline`]s.
-    ///
-    /// [`Inline`]: self::Inline
-    Multiple(NestedContent),
+    Plain(Plain),
 }
 
 impl Inline {
-    /// Create new [`Inline`] from given content depending on the given kind.
+    /// Create new [`Inline::Plain`], [`Inline::Multiple`] [`Inline::EndOfLine`] from given content
+    /// depending on the given kind.
     ///
     /// # Arguments
     ///
     /// * `content` - the [`InlineContent`] put inside the created [`Inline`]
-    /// * `kind` - the [`TokenKind`] used to define the kind of [`Inline`] that should be created
+    /// * `span` - [`Span`] that is occupied by the given content
+    /// * `kind` - the [`TokenKind`] used to choose one of the three options
     ///
     /// [`Inline`]: self::Inline
+    /// [`Inline::Plain`]: self::Inline::Plain
+    /// [`Inline::Multiple`]: self::Inline::Multiple
+    /// [`Inline::EndOfLine`]: self::Inline::EndOfLine
     /// [`TokenKind`]: crate::TokenKind
     /// [`InlineContent`]: self::content::InlineContent
-    pub fn new(content: InlineContent<PlainContent, NestedContent>, kind: TokenKind) -> Self {
-        let consume_as_plain = |content| match content {
-            InlineContent::Plain(plain_content) => Self::Plain(plain_content),
-            InlineContent::Nested(nested_content) => Self::Multiple(nested_content),
-        };
+    pub fn plain_or_eol(content: impl Into<String>, span: Span, kind: TokenKind) -> Self {
+        match kind {
+            TokenKind::EndOfLine => Self::EndOfLine(EndOfLine {
+                content: content.into(),
+                span,
+            }),
+            _ => Self::Plain(Plain {
+                content: content.into(),
+                span,
+            }),
+        }
+    }
+
+    /// Creates a new [`Inline::Plain`] from the given content with the given [`Span`].
+    pub fn plain(content: String, _kind: TokenKind, span: Span) -> Self {
```

Comment:
I think you're right. One edge case that occurred to me is, currently it is not possible to have an explicit newline in a `plain` inline. I don't know if that's relevant for LSP implementation. For example, verbatim inline would not contain any explicit newlines, because the content is simply a String...